libvast/src/meta_index.cpp

```diff
@@ -13,67 +13,194 @@
 #include "vast/logger.hpp"
 #include "vast/table_index.hpp"
 #include "vast/table_slice.hpp"
+#include "vast/time.hpp"
 
-namespace vast {
+#include "vast/detail/overload.hpp"
+#include "vast/detail/set_operations.hpp"
+#include "vast/detail/string.hpp"
 
-// -- member types -------------------------------------------------------------
+namespace vast {
 
-meta_index::interval::interval()
-  : from(timestamp::max()),
-    to(timestamp::min()) {
-  // nop
+meta_index::meta_index() : make_synopsis_{make_synopsis} {
 }
 
-meta_index::interval::interval(timestamp first, timestamp last)
-  : from(first),
-    to(last) {
-  // nop
+void meta_index::add(const uuid& partition, const table_slice& slice) {
+  auto& part_synopsis = partition_synopses_[partition];
+  auto& layout = slice.layout();
+  if (blacklisted_layouts_.count(layout) == 1)
+    return;
+  auto i = part_synopsis.find(layout);
+  table_synopsis* table_syn;
+  if (i != part_synopsis.end()) {
+    table_syn = &i->second;
+  } else {
+    // Create new synopses for a layout we haven't seen before.
+    i = part_synopsis.emplace(layout, table_synopsis{}).first;
+    table_syn = &i->second;
+    for (auto& field : layout.fields)
+      if (i->second.emplace_back(make_synopsis_(field.type)))
+        VAST_DEBUG(this, "created new synopsis structure for type", field.type);
+    // If we couldn't create a single synopsis for the layout, we will no
+    // longer attempt to create synopses in the future.
+    auto is_nullptr = [](auto& x) { return x == nullptr; };
+    if (std::all_of(table_syn->begin(), table_syn->end(), is_nullptr)) {
+      VAST_DEBUG(this, "could not create a synopsis for layout:", layout);
+      blacklisted_layouts_.insert(layout);
+    }
+  }
+  VAST_ASSERT(table_syn->size() == slice.columns());
+  for (size_t col = 0; col < slice.columns(); ++col) {
+    if (auto& syn = (*table_syn)[col])
+      for (size_t row = 0; row < slice.rows(); ++row)
+        syn->add(*slice.at(row, col));
+  }
 }
 
-// -- properties ---------------------------------------------------------------
+std::vector<uuid> meta_index::lookup(const expression& expr) const {
+  VAST_ASSERT(!caf::holds_alternative<caf::none_t>(expr));
+  // TODO: we could consider a flat_set<uuid> here, which would then have
+  // overloads for inplace intersection/union and simplify the implementation
+  // of this function a bit.
+  using result_type = std::vector<uuid>;
+  auto all_partitions = [&] {
+    result_type result;
+    result.reserve(partition_synopses_.size());
+    std::transform(partition_synopses_.begin(),
+                   partition_synopses_.end(),
+                   std::back_inserter(result),
+                   [](auto& x) { return x.first; });
+    std::sort(result.begin(), result.end());
+    return result;
+  };
+  return caf::visit(detail::overload(
+    [&](const conjunction& x) -> result_type {
+      VAST_ASSERT(!x.empty());
+      auto i = x.begin();
+      auto result = lookup(*i);
```

Comment:
Looks like potential UB if `x.begin() == x.end()`.

Reply:
For that reason I put an assertion in there. You don't think it's enough?

Reply:
Ah, sorry. Didn't spot the assertion.